packages/governance/src/tests/test_multisig.cairo

```diff
@@ -0,0 +1,1588 @@
+    start_cheat_caller_address(contract_address, ALICE());
+    state.submit_transaction(to, selector, calldata, 0);
+
+    // Confirm by Bob
+    start_cheat_caller_address(contract_address, BOB());
+    state.confirm_transaction(id);
+    assert_eq!(state.is_confirmed_by(id, BOB()), true);
+    assert_eq!(state.get_transaction_confirmations(id), 1);
+
+    // Try to confirm again by Bob
+    state.confirm_transaction(id);
+}
+
+//
+// Revoke confirmation
+//
+
+#[test]
+fn test_revoke_confirmation() {
+    let (quorum, signers) = DEFAULT_DATA();
+    let mut state = setup_component(quorum, signers);
+    let mut spy = spy_events();
+    let contract_address = test_address();
+
+    // Submit by Alice
+    let Call { to, selector, calldata } = build_call(MockCall::AddNumber(42));
+    start_cheat_caller_address(contract_address, ALICE());
+    let id = state.submit_transaction(to, selector, calldata, 0);
+
+    // Confirm by Bob
+    start_cheat_caller_address(contract_address, BOB());
+    state.confirm_transaction(id);
+
+    // Confirm by Charlie
+    start_cheat_caller_address(contract_address, CHARLIE());
+    state.confirm_transaction(id);
+
+    // Revoke confirmation by Charlie
+    spy.drop_all_events();
+    assert_tx_state(id, TransactionState::Confirmed);
+    assert_eq!(state.is_confirmed_by(id, CHARLIE()), true);
+    assert_eq!(state.get_transaction_confirmations(id), 2);
+    state.revoke_confirmation(id);
+    assert_tx_state(id, TransactionState::Pending);
+    assert_eq!(state.is_confirmed_by(id, CHARLIE()), false);
+    assert_eq!(state.get_transaction_confirmations(id), 1);
+    spy.assert_only_event_confirmation_revoked(contract_address, id, CHARLIE());
+}
+
+#[test]
+fn test_tx_not_confirmed_after_signer_removal() {
+    let (quorum, signers) = DEFAULT_DATA();
+    let mut state = setup_component(quorum, signers);
+    let contract_address = test_address();
+
+    // Submit & confirm by Alice
+    let Call { to, selector, calldata } = build_call(MockCall::AddNumber(42));
+    start_cheat_caller_address(contract_address, ALICE());
+    let id = state.submit_transaction(to, selector, calldata, 0);
+    state.confirm_transaction(id);
+
+    // Confirm by Bob
+    start_cheat_caller_address(contract_address, BOB());
+    state.confirm_transaction(id);
+
+    // Check state before removal
+    assert_tx_state(id, TransactionState::Confirmed);
+    assert_eq!(state.is_confirmed_by(id, BOB()), true);
+    assert_eq!(state.get_transaction_confirmations(id), 2);
+
+    // Remove Bob from signers
+    start_cheat_caller_address(contract_address, contract_address);
+    state.remove_signers(quorum, array![BOB()].span());
+
+    // Check state after removal
+    assert_tx_state(id, TransactionState::Pending);
+    assert_eq!(state.is_confirmed_by(id, BOB()), true);
+    assert_eq!(state.get_transaction_confirmations(id), 1);
+}
+
+#[test]
+fn test_can_revoke_confirmation_after_being_removed() {
```

Comment:
I think it's the way it should be and I'm looking at that from the following perspective:
- If Bob confirmed a tx when he was a signer, then `is_confirmed_by` should return `true` , even if he was removed from signers. This is a tricky case, but returning `false` is even more misleading
- On the other hand, his confirmation no longer counts to the quorum, since Bob isn't considered signer now. Bob isn't present in the list of signers, so  `is_confirmed_by(Bob)` can be called only if someone is specifically interested if Bob has confirmed the tx
- This logic must be consistent for any tx regardless of its status (Pending, Confirmed or Executed)

There's an important point to note: We _could_ change the logic of `is_confirmed_by(Bob, tx)` so that it returns `true` only when Bob is a signer AND he has confirmed the `tx`. But it would introduce problems for already executed txs that were confirmed by Bob (Bob was removed from signers after tx execution):
- tx's confirmation count would be less than `quorum` 
- there would be no confirmation from Bob

Reply:
I agree with the @immrsd approach, as long as the confirmation doesn't count for quorum I don't think we should revoke it when the signer is removed.

Reply:
All right. Let's make sure we document this possibility so when users integrate the component, they're aware that `is_confirmed_by` should not be relied upon for the current state of the quorum

> We could change the logic of is_confirmed_by(Bob, tx) so that it returns true only when Bob is a signer AND he has confirmed the tx. But it would introduce problems for already executed txs that were confirmed by Bob

Yeah I agree this wouldn't be a great approach